src/algorithms/highlevel/crosssimilaritymatrix.cpp

```diff
@@ -0,0 +1,233 @@
+query and reference song.\n" "\n"
+"\n\n"
+"References \n"
+"----------"
+"\n\n"
+"[1]. Serra, J., Gómez, E., & Herrera, P. (2008). Transposing chroma representations to a common key, IEEE Conference on The Use of Symbols to Represent Music and Multimedia Objects. \n"
+"[2]. Tralie, C.J., 2017. Geometric Multimedia Time Series (Doctoral dissertation, Duke University)\n"
+"[3]. Serra, J., Serra, X., & Andrzejak, R. G. (2009). Cross recurrence quantification for cover song identification.New Journal of Physics.\n"
+"[4]. Serra, Joan, et al. Chroma binary similarity and local alignment applied to cover song identification. IEEE Transactions on Audio, Speech, and Language Processing 16.6 (2008).\n"
+"\n");
+
+
+void CrossSimilarityMatrix::configure() {
+  _tau = parameter("tau").toInt();
+  _m = parameter("m").toInt();
+  _kappa = parameter("kappa").toDouble();
+  _noti = parameter("noti").toInt();
+  _oti = parameter("oti").toBool();
+  _toBlocked = parameter("toBlocked").toBool();
+}
+
+void CrossSimilarityMatrix::compute() {
+    // get inputs and output
+    std::vector<std::vector<Real> > queryFeature = _queryFeature.get();
+    std::vector<std::vector<Real> > referenceFeature = _referenceFeature.get();
+    std::vector<std::vector<Real> >& csm = _csm.get();
+    std::vector<std::vector<Real> > pdistances;
+
+    // check whether to transpose by oti
+    if (_oti == true) {
+        int otiIdx = optimalTranspositionIndex(queryFeature, referenceFeature, _noti);
+        std::rotate(referenceFeature.begin(), referenceFeature.end() - otiIdx, referenceFeature.end());
+    }
+
+    // check if delay embedding needed
+    if (_toBlocked == true) {
+        // construct time embedding from input chroma features
+        std::vector<std::vector<Real> >  timeEmbedA = toTimeEmbedding(queryFeature, _m, _tau);
+        std::vector<std::vector<Real> >  timeEmbedB = toTimeEmbedding(referenceFeature, _m, _tau);
+        // pairwise euclidean distance
+        pdistances = pairwiseDistance(timeEmbedA, timeEmbedB);
+
+        // free memorry of input feature arrays
+        std::vector<std::vector<Real> >().swap(timeEmbedA);
+        std::vector<std::vector<Real> >().swap(timeEmbedB);
+    }
+    else {
+        // pairwise euclidean distance
+        pdistances = pairwiseDistance(queryFeature, referenceFeature);
+
+        // free memorry of input feature arrays
+        std::vector<std::vector<Real> >().swap(queryFeature);
+        std::vector<std::vector<Real> >().swap(referenceFeature);
+    }
+
+    if (pdistances.empty())
+       throw EssentiaException("empty array found while calculating euclidian similarity");
+
+    // transposing the array of pairwsie distance
+    std::vector<std::vector<Real> > tpDistances = transpose(pdistances);
+
+    std::vector<std::vector<Real> > ephX(pdistances.size());
+    std::vector<std::vector<Real> > ephY(tpDistances.size());
+
+    std::vector<Real> tempXrow;
```

Comment:
Using tempXrow appears redundant as it's size is always 1. Populate ``ephX`` directly?